text/3440-virt-self.md

```diff
@@ -0,0 +1,189 @@
+
+Of course, this behaviour may be what is desired. But sometimes, virtualizing at depth is more appropriate. Certainly when adopting OOP patterns.
+
+When I first stumbled upon this, I wondered if I could simply take `self` as `&dyn System` (or `Arc` or `Box` equivalent, etc.). Nope.
+
+The code I was working on required me to keep the trait `System` as one unit so I did not investigate possible ways to split it up.
+
+What I came up is passing a &dyn System virtualized self so now System looked like this
+
+```rust
+pub trait System {
+    fn op1(&self, vself: &dyn System) -> SystemOp1;
+    fn op2(&self, vself: &dyn System) -> SystemOp2;
+}
+impl System for SystemA {
+    fn op1(&self, vself: &dyn System) -> SystemOp1 {
+        vself.op2() * 5
+    }
+    fn op2(&self, vself: &dyn System) -> SystemOp2 {
+        3
+    }
+}
+```
+
+And then calling `op2` with essentially another copy of `self`:
+```rust
+let system = DoubleOp2<SystemA> {};
+system.op2(&system2); // works!
+```
+
+Of course this design is a bit clunky, but is very powerful. It allows complete control over what is virtualized and what isn't at any point in the call chain for every implementation.
+
+As I have been thinking about this more and more, I wondered what Rust with this in the language would look like. This RFC proposes a sample syntax (bikeshedding welcome, but please focus on the concepts).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Essentially, we now allow `&virt self` as a shorthand for the above.
+
+```rust
+pub trait System {
+    fn op1(&virt self) -> SystemOp1;
+    fn op2(&virt self) -> SystemOp2;
+}
+```
+
+The effect is that rust will manage two pointers - one concretely typed and one dynamic. The The concretely typed (i.e. `self`) follows the existing Rust rules. The dynamic one is either `self` as `dyn System` again or an existing dynamic `self` (this is like choosing `self` or `vself` in the previous section).
+
+Syntax could be e.g.
+```rust
+impl System for SystemA {
+    fn op1(&virt self) -> SystemOp1 {
+        // alternative 1
+        virt(self).op2() * 5
+        // alternative 2, I like this one most because it reminds me of C
+        self->op2() * 5
+        // something else...
+    }
+    fn op2(&virt self) -> SystemOp2 {
+        3
+    }
+}
+```
+
+Working with the second syntax, the difference between `self.op2()` and `self->op2()` would be static vs dynamic dispatch. In other words, it would allow us to call `DoubleOp2<SystemA>::op2()` from within `SystemA::op1()` when the call to it is made from `DoubleOp2<SystemA>::op1()`.
+
+If the called method is also declared virt, using `self->op2()` will retain the `vself` the same as originally passed to `op1`. Otherwise, it will replace it by `self as &dyn System`.
+
+Outside of traits, `a->b()` wouldn't compile. 
```

Comment:
But macros that use this syntax would be meant for the trait case, otherwise normal syntax will still work as it does today. So what is really lost?

Reply:
Currently, in syntax generation through macros, an expression is an expression and you don't have to think about it beyond that. This would create a "subclass" of `$trait_expr` that could not be used where all `$expr` could be used. This is a severe problem for macros concealing their implementation details:

Now macros that pass in an expression using this syntax to the inside of a trait and use that to do some work break if they instead pass the expression through to inside free functions or closures to do that work. In order to create slightly more opacity for traits, it blows a giant hole through opacity for macros.

Reply:
( maybe that's a bit strong, but "expressions are expressions" is kiiinda important in a general sense. )